### PR TITLE
Added query params functionality to Cursor.execute via PREPARE statement

### PR DESCRIPTION
**Pull request by harph**

### CODE REVIEW

File: presto/dbapi.py
Comment (hashhar) [inline presto/dbapi.py:22]: Since we are using `?` as the parameter style, we should add a module level variable `paramstyle` set to `?` to ensure we comply with the DB-API (since most tools assume the default to be `%(name)` style parameters).

See https://www.python.org/dev/peps/pep-0249/#paramstyle

**Comment by hashhar** [comment]
@harph The code solves the core problem correctly and the ideas look good.

But this might need some changes around usability (prepared statements are not re-used nor can the user try to re-use them, signature change of public methods allowing to pass headers etc.) and DB-API compliance.

I'll follow up with a detailed review in a while.

File: presto/dbapi.py
Comment (hashhar) [inline presto/dbapi.py:290]: This method implicitly plays with the string representation of `params` which can cause issues (large numbers which python prints using scientific notation, arrays etc.).

File: presto/dbapi.py
Comment (hashhar) [inline presto/dbapi.py:248]: Maybe rename to `prepare_statement` since we don't actually `EXECUTE` here.

File: presto/dbapi.py
Comment (hashhar) [inline presto/dbapi.py:319]: @findepi @electrum Should we start looking at how other DB-API drivers in Python solve this by providing multiple Cursor classes and over-ride the `execute()` and related functions accordingly in the sub-classes?

Will keep things cleaner and allow adding new Cursors that can maybe return the data as list of dicts etc.

**Comment by hashhar** [comment]
In my opinion, we should consider adding a new Cursor subclass and the `connect()` method can take an argument to return the correct Cursor class.
This will allow the `PreparedCursor` class to cache prepared statements and transparently re-use them (see https://github.com/mysql/mysql-connector-python/blob/master/lib/mysql/connector/cursor.py#L1089 for the idea).

What I am aiming for is to allow the user control over the preparing, execution and deallocation of statements. Right now all of them are coupled within the `execute()` method.

**Comment by findepi** [comment]
> This will allow the PreparedCursor class to cache prepared statements and transparently re-use them

I think this is not as important for Presto as it may be for  eg MySQL.
MySQL is specializing in low latency queries, so caching is more important for them.

Also, Presto re-plans prepared staement every time currently. Idk whether MySQL does the same.

> What I am aiming for is to allow the user control over the preparing, execution and de-allocation of statements. Right now all of them are coupled within the execute() method.

More control is also more complexity (for users, for maintainers).
I wouldn't "offer" more complexity to users unless this is really needed.

> Another possibility is to merge and take this in a follow up PR (risky since after release people will start depending on the current behaviour).

 what is the end-user visible behavior users could become attached to?

**Comment by hashhar** [comment]
> Also, Presto re-plans prepared staement every time currently. Idk whether MySQL does the same.

Then I think we can go with the current implementation since the benefit I was thinking of of re-using prepared statements doesn't really exist.

> More control is also more complexity (for users, for maintainers).
> I wouldn't "offer" more complexity to users unless this is really needed.

Same as above since the control I was thinking about was something like a user can prepare 2 queries ahead of time and fire them within a loop based on some condition. I wanted to change:

```
for i in 1 to 10:
  if i % 2:
    s1 := PREPARE
    EXECUTE(s1)
    DEALLOCATE(s1)
  else:
    s2 := PREPARE
    EXECUTE(s2)
    DEALLOCATE(s2)
```
to 

```
s1 := PREPARE
s2 := PREPARE
for i in 1 to 10:
  if i % 2:
    EXECUTE(s1)
  else:
    EXECUTE(s2)

DEALLOCATE(s1)
DEALLOCATE(s2)
```
> > Another possibility is to merge and take this in a follow up PR (risky since after release people will start depending on the current behaviour).

The `additional_headers` argument that the `client.execute()` method now takes. I can imagine people using that to pass authentication headers/custom headers on a per-query basis to Presto.

With the above clarification around re-planning the query and no benefit to caching the PREPAREd statement I think the only thing left to discuss is the `additional_headers` on the client.

**Comment by findepi** [comment]
> The additional_headers argument that the client.execute() method now takes.

Can we avoid this?
Ideally we keep the user-facing code surface minimal, for better future of our selves :)

**Comment by hashhar** [comment]
> > The additional_headers argument that the client.execute() method now takes.
> 
> Can we avoid this?
> Ideally we keep the user-facing code surface minimal, for better future of our selves :)

One way could be model it similar to transactions? We add an attribute (dict) to `PrestoRequest` to track the prepared statement name to the value of `X-Presto-Added-Prepare`.

The flow can then be:

1. User calls `cursor.execute(sql, params)`
2. If `params` are present, generate a `statement_name` and call `cursor._prepare_statement(sql, statement_name)`
3. In `cursor._prepare_statement(sql, statement_name)`:
    - Create a copy of `self._request` (ctor vs deepcopy?)
    - Execute `PREPARE {statement_name} FROM {sql}` using the copied `request`
    - Create a new `self._request` via constructor and pass in the `X-Presto-Added-Prepare` response value as `prepared_statement` to PrestoRequest
    - Return the new `request` object that can be used to execute queries with the prepared statement (similar to Transactions)
4. Call `cursor._execute_prepare_statement(sql, request, params)` which execute `EXECUTE {request.prepared_statement_name} USING {params}` using the new `request`
5. Call `cursor._deallocate_prepare(request)` which execute `DEALLOCATE PREPARE {request.prepared_statement_name}`

Another option could be to wrap the behaviour from 3 to 5 inside a `PreparedStatement` class. Helps keep things compartmentalized.

In this scenario the only user-visible change would be the new `params` property to `cursor.execute()`.

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py:22]: @hashhar Are you suggesting adding `paramstyle = '?'` as a  module variable/constant so that people can reference it when writing queries?

Comment (hashhar) [reply presto/dbapi.py:22]: Yes. Both for users writing queries and external tools that generate queries (I've seen some for PostgreSQL).

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py:290]: Good point. What do you suggest in this case?

I think it's fair to assume that any complex object in the `params` the  must implement the `__str__` function  in order to be treated properly. Otherwise, they should be parsed/converted on the user's implementation before appending them to the `params`.

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py:248]: Sounds good. I'll do the same to with the deallocate part of it.

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py:22]: Perfect. I'll make that change.

**Comment by findepi** [comment]
> In this scenario the only user-visible change would be the new `params` property to `cursor.execute()`.

that's the goal, so i'm fine with this :)


> The additional_headers argument that the client.execute() method now takes

i mis-read this perviously, sorry. I hope we can consider `client` as the impl details of the module, 
so we do not expose new contract fo users here.

Is my assumption now sound?

File: presto/dbapi.py
Comment (hashhar) [reply presto/dbapi.py:290]: @harph You are correct. I think we can leave as is in that case but clarify this in the docstring of the method to avoid people like me being caught by surprise.

**That still leaves one concern and that is regarding `params` whose values themselves include a `,`.**

Been a little long since I used Python with any DB. Almost every DB interprets the values passes as parameters as `LITERALS` by converting to a string.

**Comment by hashhar** [comment]
@findepi Yes, if we consider `client` to be internal (the module docstring makes it **sound** public) then the PR can be merged once the `paramstyle`, tests for params with `,` and new method names are renamed.

Sorry for the unnecessary theoretical journey in that case. :sweat_smile:

File: presto/dbapi.py
Comment (hashhar) [reply presto/dbapi.py:290]: https://github.com/prestosql/presto-python-client/pull/16/files#diff-e6510cf8abf3e99495eee2b0d39cd839R54 might be useful.

@harph

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py:286]: good catch. I'll remove it

File: presto/dbapi.py
Comment (findepi) [reply presto/dbapi.py:290]: >I think it's fair to assume that any complex object in the params the must implement the __str__ function in order to be treated properly.

__str__ has different contract and different requirements than Presto's parser.

I'd rather err on the side of not supporting a type, rather than call str on something I do not know.

I assume that supporting totally custom types is not a requirement, so we do not have to worry about them.

File: presto/exceptions.py
Comment (losipiuk) [inline presto/exceptions.py:221]: add newline

File: presto/dbapi.py
Comment (losipiuk) [inline presto/dbapi.py:348]: shouldn't we wrap section above in `try:` and execute `_execute_deallocate_prepare` also for flow with exeption in `finally:`?

File: presto/dbapi.py
Comment (losipiuk) [inline presto/dbapi.py:325]: instead having a comment here I would suggest extracting `_generate_unique_statement_name` function. Calling it will be self explanatory.

File: presto/dbapi.py
Comment (losipiuk) [inline presto/dbapi.py:340]: I was thinking that maybe we should change headers map in  `self._query`. But passing it explicitly here seems fine too.

File: tests/test_client.py
Comment (losipiuk) [inline tests/test_client.py:691]: new line

File: presto/dbapi.py
Comment (losipiuk) [inline presto/dbapi.py:271]: Can't we use `self._create_request()` here?
Is it important to call `PREPARE` using request containing information regarding current transaction?

File: tests/test_client.py
Comment (harph) [reply tests/test_client.py]: done

File: presto/exceptions.py
Comment (harph) [reply presto/exceptions.py]: done

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py:271]: The reason why I decided to do it this way is that the `Cursor` class receives a request object which can technically be different from what `_create_request` originally returns, if we assume that `Cursor` class can be instantiated directly. Also `Cursor` does no have access to the `_create_request` function since that one belongs to the `Connection` class.

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py:348]: Good point. I'll do that.

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py]: Addressed.

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py]: done

File: presto/client.py
Comment (harph) [reply presto/client.py:568]: That does not apply to `post`, `get`, and `delete` because they belong to the `PrestoRequest` instead of the `PrestoQuery` class. Did you mean `execute` and `cancel` instead?

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py:22]: Done

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py]: Done

**Comment by harph** [comment]
@hashhar @findepi Just to confirm, are we ok with leaving `addtional_headers` as parameter of the `client.execute` function given that is consider part of the module implementation?

**Comment by findepi** [comment]
> @hashhar @findepi Just to confirm, are we ok with leaving `addtional_headers` as parameter of the `client.execute` function given that is consider part of the module implementation?

i am ok

File: presto/client.py
Comment (hashhar) [reply presto/client.py:568]: You are correct. I should've said `execute` and `cancel`. I checked the headers returned on `execute` and `cancel` but they didn't look useful enough generally (yet) so we can let this go.

File: presto/dbapi.py
Comment (hashhar) [inline presto/dbapi.py:338]: nit: typo

```suggestion
                # as it will be returned by the function
```

**Review by hashhar — APPROVED** [review]
Looks good to go % the discussion about `_str_` vs what was proposed in [ParamEscaper](https://github.com/prestosql/presto-python-client/pull/16/files#diff-e6510cf8abf3e99495eee2b0d39cd839R49) in another PR.

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py]: Fixed

File: presto/dbapi.py
Comment (harph) [reply presto/dbapi.py]: Done

**Comment by harph** [comment]
What's the plan for merging and releasing this?

**Comment by findepi** [comment]
@hashhar any other comments?
@losipiuk do you want to take a look?

otherwise I can merge & release

**Comment by hashhar** [comment]
> Using `_str_` vs what was proposed in [ParamEscaper](https://github.com/prestosql/presto-python-client/pull/16/files#diff-e6510cf8abf3e99495eee2b0d39cd839R49) in another PR.

@findepi @harph should this be a follow up PR?

**Comment by findepi** [comment]
> Converting params by calling str vs making the code type-aware and type-safe.
> 
> @findepi @harph should this be a follow up PR?

i'm fine, please make sure we have issues covering this

**Comment by harph** [comment]
> > Using `_str_` vs what was proposed in [ParamEscaper](https://github.com/prestosql/presto-python-client/pull/16/files#diff-e6510cf8abf3e99495eee2b0d39cd839R49) in another PR.
> 
> Converting params by calling str vs making the code type-aware and type-safe.
> 
> @findepi @harph should this be a follow up PR?

I'm ok addressing it on this one.

**Comment by losipiuk** [comment]
I am fine with merging this as is and do improvements in followup PR.

**Review by losipiuk — APPROVED** [review]

**Review by hashhar — APPROVED** [review]
LGTM :shipit:

**Comment by findepi** [comment]
Released in `0.301.0`

**Comment by ps-cbsantos** [comment]
Hi all,

I'm trying to use %s as the placeholder to my stmt, but always getting the error bellow:
`TrinoUserError(type=USER_ERROR, name=SYNTAX_ERROR, message="line 1:130: mismatched input '='. Expecting: <predicate>", query_id=20211221_205415_00100_hau44)`

The query I'm trying to run:
`cursor.execute("select  * from system.runtime.nodes where status=%s", ['active'])`

I followed the ways as described in the [test](https://github.com/trinodb/trino-python-client/pull/16/files#diff-e6510cf8abf3e99495eee2b0d39cd839R49), but could not get it done.

**Comment by harph** [comment]
@ps-cbsantos take a look at this test that show you how it works: https://github.com/trinodb/trino-python-client/blob/a7438556213bcf954f8a7a5063c7823bbb177edb/integration_tests/test_dbapi.py#L73

You basically have to replace `%s` by `?`

**Pull request merged**
